### PR TITLE
[RFC] Proof of concept of loading trained model

### PR DESCRIPTION
**Pull request by simlmx**
Not meant to merge but as a conversation starter!

* Add fixtures for PV and NWP data
* Add fixture serialized model from `pv-site-prediction`, which in turn is added as a submodule
* Standardize the interfaces to support both serialized models and non-trained models like the `cosine` model.
* Add fixture configs for both models (cosine and serialized) and run `app.py` for each.

TODOs
- [ ] Edit this description
- [ ] Make the github tests pass
- [ ] Clean the commits (worse case squash everything)
- [ ] Self-Review

### CODE REVIEW

File: pv_site_production/app.py [hunk 1]
```diff
@@ -1,42 +1,77 @@
...
-    Make app method
+    """Make app method"""
+    with open(config_path) as f:
+        config = yaml.safe_load(f)
 
-    :param model_name: the model name, current is only 'cos'
-    :return:
-    """
+    run_model_func = config["run_model_func"]
+    module, func_name = run_model_func.rsplit(".", maxsplit=1)
```
Comment (peterdudfield) [inline pv_site_production/app.py:59]: should this just be two different ones in the configruation

**Comment by peterdudfield** [comment]
Working Progress comments

- I like the use of config files
- And good use of peotry (keen to perhaps put this in a seperate PR), but I think thats what you plan to do.
- Bit disappointed that `ocf_datapipes` has been stripped out. I _really_ want this to be used. This will keep all the `Database` connections and all the other fun things that we should not worry about here. We've learnt in the passed it has been really useful to strip out the dataloading part. This means other modules / forecast can really use lots of things together. Let's definately discuss this, but right now, I'll be disappointed if we dont use it. 
- For NWP test data, I found its better to use a text fixture, you can see this in #5 . I think this keeps things tidy compared to committing files

File: pv_site_production/app.py [hunk 2]
```diff
@@ -1,42 +1,77 @@
...
-    """
-    Make app method
+    """Make app method"""
+    with open(config_path) as f:
+        config = yaml.safe_load(f)
 
-    :param model_name: the model name, current is only 'cos'
-    :return:
-    """
+    run_model_func = config["run_model_func"]
```
Comment (simlmx) [inline pv_site_production/app.py:58]:
```suggestion
    run_model_func = config["get_model_func"]
```

File: pv_site_production/app.py [hunk 3]
```diff
@@ -1,42 +1,77 @@
...
+
+    pv_ids = pv_data.list_pv_ids()
 
-    # choose model
-    if model_name == "cos":
-        run_model = run_cos_model
-    else:
-        Exception(f"Could not find model {model_name}")
+    # TODO support passing a different date.
+    now = datetime.now()
```
Comment (simlmx) [inline pv_site_production/app.py:70]:
```suggestion
    now = datetime.utcnow()
```

File: pv_site_production/app.py [hunk 4]
```diff
@@ -2,14 +2,126 @@
...
 from psp.ml.models.base import PvSiteModel
 from psp.ml.typings import PvId, Timestamp, X
 
+# Meta keys that are stil taken from our metadata file.
+# TODO Those should move to the database.
+META_FILE_KEYS = ["tilt", "orientation", "factor"]
+META_DB_KEYS = ["longitude", "latitude"]
+
+
+# TODO This probably goes to another file
```
Comment (peterdudfield) [inline pv_site_production/app.py:27]: yea, maybe not in app

File: pv_site_production/app.py [hunk 5]
```diff
@@ -40,6 +152,19 @@ def apply_model(model: PvSiteModel, pv_ids: list[PvId], ts: Timestamp) -> pd.Dat
     return pd.DataFrame.from_records(records)
 
 
+def import_from_module(module_path: str) -> Any:
```
Comment (peterdudfield) [inline pv_site_production/app.py:155]: move to utils?

File: pv_site_production/app.py [hunk 6]
```diff
@@ -55,19 +180,23 @@ def run(
...
-    pv_data = NetcdfPvDataSource(config["pv_path"])
+    session = pv_db_connection.get_session()
+    pv_systems = get_pv_systems(session=session, provider=SHEFFIELD)
 
-    pv_ids = pv_data.list_pv_ids()
+    pv_ids = [p.pv_system_id for p in pv_systems]
 
     # TODO support passing a different date.
-    now = datetime.now()
+    now = datetime(2022, 1, 1, 6)
```
Comment (peterdudfield) [inline pv_site_production/app.py:199]: maybe as click argument?

File: pv_site_production/app.py [hunk 7]
```diff
@@ -2,14 +2,126 @@
...
+
+        # TODO The fact that we have to check for two types tells me something does not get checked
+        # properly somewhere!
+        if isinstance(pv_ids, (PvId, np.integer)):
+            was_1d = True
+            pv_ids = [pv_ids]
+        else:
+            was_1d = False
+
+        pv_yields = get_pv_yield(
```
Comment (peterdudfield) [inline pv_site_production/app.py:54]: nice, thanks for using the `get_pv_yield` function

File: pv_site_production/app.py [hunk 8]
```diff
@@ -2,14 +2,126 @@
...
+        da = da.assign_coords(
+            {
+                key: ("pv_id", [meta[pv_id][key] for pv_id in pv_ids])
+                for key in META_FILE_KEYS + META_DB_KEYS
+            }
+        )
+
+        if was_1d:
+            da = da.isel(pv_id=0)
+
```
Comment (peterdudfield) [inline pv_site_production/app.py:109]: I think this is all fine, until it dosnt work, becasue data is delayed e.t.c. Thats where using the `ocf_datapipes` will be handy. 

Of course you still perhaps need to rename variables how you wish. 

Perhaps I could add a filter for `pv_ids` into the `ocf_datapipes` module so you have that functionality

File: Makefile [hunk 9]
```diff
@@ -1,12 +1,31 @@
...
+		postgres:13-alpine
+
+
 .PHONY: test
-test:
-	docker stop $(docker ps -a -q); true
-	docker-compose -f infrastructure/test-docker-compose.yml build
-	docker-compose -f infrastructure/test-docker-compose.yml run tests
+test: test-db
+	poetry run pytest tests $(ARGS)
```
Comment (peterdudfield) [inline Makefile:26]: if this works, then this is super!

File: pv_site_production/app.py [hunk 10]
```diff
@@ -1,44 +1,104 @@
...
-    # TODO
-    print(results_df)
+    _log.info("Applying model")
+    # TODO Calling/printing like this to debug but we could call `apply_model` once with all the pv
+    # ids.
+    for pv_id in pv_ids:
+        results_df = apply_model(model, pv_ids=[pv_id], ts=timestamp)
+        print(pv_id)
+        for _, row in results_df.iterrows():
+            print(f"{row['target_datetime_utc']}: {row['forecast_kw']}")
```
Comment (peterdudfield) [inline pv_site_production/app.py:102]: can you leabe the # save model comment in, just so we know where the model is saved. 

I think its nice, if we build up all the results and then pass it though once into a save function

Comment (simlmx) [reply pv_site_production/app.py:102]: But we're not saving the model here. Or you mean saving the results, eventually to the database?

**Comment by simlmx** [comment]
Closing in favor of #15

**Pull request closed without merging**